greenmail-core/src/test/java/com/icegreen/greenmail/test/ImapServerTest.java

```diff
@@ -714,4 +719,73 @@ public void messagesAdded(MessageCountEvent e) {
             store.close();
         }
     }
+
+    @Test
+    public void testMessageChangedListener() throws MessagingException {
+        CountDownLatch latch = new CountDownLatch(1);
+
+        greenMail.setUser("foo@localhost", "pwd");
+
+        final IMAPStore store = greenMail.getImap().createStore();
+        store.connect("foo@localhost", "pwd");
+
+        GreenMailUtil.sendTextEmail("foo@localhost", "bar@localhost", "Test subject", "Test message",
+                ServerSetupTest.SMTP);
+
+        Thread thread = null;
+        try {
+            Folder inboxFolder = store.getFolder("INBOX");
+            inboxFolder.open(Folder.READ_ONLY);
+            thread = new Thread(() -> {
+                try {
+                    while (!Thread.currentThread().isInterrupted()) {
+                        ((IMAPFolder) inboxFolder).idle(true);
+                    }
+                } catch (MessagingException ex) {
+                    assertThat(false).isTrue();
+                }
+            });
+            thread.start();
+
+            int[] messages = new int[] { 0 };
+            MessageChangedListener listener = new MessageChangedListener() {
+                @Override
+                public void messageChanged(MessageChangedEvent e) {
+                    assert e.getMessageChangeType() == MessageChangedEvent.FLAGS_CHANGED;
+                    messages[0] = e.getMessage().getMessageNumber();
+                    latch.countDown();
+                }
+            };
+            inboxFolder.addMessageChangedListener(listener);
+
+            new Thread(() -> {
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException e1) {
+                    // Ignore
+                }
+
+                try {
```

Comment:
As commented, using the internal (server-side) API won't trigger any listeners. Because you directly manipulate the message in memory, workaround any JavaMail listener mechanism.

Instead, use IMAP / JavaMail:

```inboxFolder.getMessages()[0].setFlag(DELETED, true); // Triggers the listener, as this results in an IMAP command```

Reply:
@marcelmay I added another commit that does what you suggest. First I try with the same variable on which I've added the listener - that works as expected. Then I try with a new variable - then it doesn't. That's what I meant before by "external" event. Sorry, I should have removed the "server-side" setting of flags before. But please check the test now.